plugins/hypervisors/xenserver/src/com/cloud/hypervisor/XenServerGuru.java

```diff
@@ -167,35 +170,62 @@ public boolean trackVmHostChange() {
 
     @Override
     public Pair<Boolean, Long> getCommandHostDelegation(long hostId, Command cmd) {
```

Comment:
in terms of cyclic complexity this method doesn't improve

Reply:
Actually, it depends on how you count. Normally you would need to build a decision graph to calculate. However, the most import thing in my opinion is the removal of 3/4 conditionals one inside the other. And More specifically this block:
```
if (hypervisorVersion != null && !hypervisorVersion.equalsIgnoreCase("6.1.0")) {
-                        if (!(hypervisorVersion.equalsIgnoreCase("6.2.0") &&
-                                !(snapshotHotFixVersion != null && snapshotHotFixVersion.equalsIgnoreCase(XenserverConfigs.XSHotFix62ESP1004)))) {
-                            return new Pair<Boolean, Long>(Boolean.TRUE, new Long(host.getId()));
-                        }
-                    }
````

Those negations altogether do not compute well in my mind...

Reply:
all the conditional returns compensate more then fully the complexity of the nested conditionals, so i think the complexity grew. But of course you are right and we should actually measure (lazy mode on - quickly)

Reply:
Yes you are right!
I have seen people counting the return as a point for the final metric; but, I am not sure yet if they count or not. In my view the decision point is in the conditional that enabled the return to be executed and not in the return per se.